### PR TITLE
Refactor MPI_Wait/MPI_Test implementations

### PR DESCRIPTION
**Pull request by hajimefu**
# Introduction

This is a proposal to refactor `MPI_Wait` family (Wait/Waitall/Waitany/Waitsome) and `MPI_Test` family (Test/Testall/Testany/Testsome) implementations so an MPICH device can override them. By  doing so, a device may for example implement more fine-grain wait on a particular request by polling a limited set of hardware context.

The key idea is that `MPI_Wait*/Test*` call `MPID_Wait*/Test*`, which a device defines. If the device doesn't want to bother implementing its own, it may simply call `MPIR_Wait*_impl/Test*_impl` as a fallback.

Following a standard convention, `MPID_Wait*/Test*` take `MPIR_Request *` **object pointers** instead of `MPI_Request` **handles**. This requires refactoring of existing `MPIR_Wait*_impl/Test*_impl` as well, as they currently take handles.

Replaces #2708 

# Change Summary

This PR essentially includes the following changes:
* Refactors `MPIR_Wait*_impl/Test*_impl`.
    * Change argument type from request handle to request object pointer
    * They are now responsible only for making a progress and checks for completion
    * They used to translate handles, free the request object after completion, nullify the handle, and set the user status. Now these are caller's responsibility except for a very limited situation (e.g. generalized request.)
* Modifies `MPI_Wait*/Test*` so they call `MPID_Wait*/Test*`
* Introduce `#define` macro definitions in CH3/CH4 that redirect `MPID_Wait*/Test*` to `MPIR_Wait*_impl/Test*_impl`

# Potential Drawbacks

Especially for calls that handle multiple requests at once (e.g. `MPI_Waitall`), the `impl` functions typically used to process requests in a single loop, in which everything including handle translation, progress, object deallocation etc. happened. With this PR we had to split this into multiple loops in different functions, there may be performance overhead. Hope is that more optimized device-specific implementation could offset it somehow.

Here are some results from osu_mbr_mr message rate with CH4 on KNL machines (Intel(R) Xeon Phi(TM) CPU 7210 @ 1.30GHz) with OPA-1. osu_mbr_mr uses `MPI_Waitall` to wait for `Isend/Irecv` to complete. Since `MPI_Waitall` has an optimized path when `MPI_STATUES_IGNORE` is used, I also modified the benchmark to specify `MPI_STATUES_IGNORE` (the original code wasn't using it.) Results for 8-byte messages are reported, and these numbers are averages of 10 runs.

## Results with original osu_mbr_mr (no `MPI_STATUSES_IGNORE`)
```
window size  64    8192
------------------------
master   916535  773122
this PR  910998  739364
```

## Results with modified osu_mbr_mr (with `MPI_STATUSES_IGNORE`)
```
window size  64    8192
-----------------------
master   940524  802396
this PR	 935359  775312
```

## Detailed Configurations

* Compiler: ICC 17.0.4
* MPICH configuration
```
  $ /home/hajimefu/src/mpich-review/configure -C --prefix=/home/hajimefu/opt/mpich-psm2-yes-tg-intel --disable-perftest --with-libfabric=/home/hajimefu/opt/libfabric-psm2 --disable-ft-tests --with-fwrapname=mpigf --with-filesystem=ufs+nfs --enable-timer-type=linux86_cycle --enable-romio --with-mpe=no --with-smpcoll=yes --with-assert-level=0 --enable-shared --enable-static --enable-error-messages=yes --enable-visibility --enable-large-tests --enable-g=none --enable-timing=none --enable-error-checking=no --enable-fast=all,O3 --disable-debuginfo --with-device=ch4:ofi:psm2 --enable-handle-allocation=default --enable-threads=multiple --without-valgrind --enable-timing=none --enable-ch4-shm=yes --enable-thread-cs=global --with-ch4-netmod-ofi-args= MPICHLIB_CFLAGS=-O3 -Wall -ggdb  -msse2 -msse4.2 -mcrc32 -mavx512f -ggdb -Wall -mtune=generic -std=gnu99 MPICHLIB_CXXFLAGS=-O3 -Wall -ggdb  -msse2 -msse4.2 -mcrc32 -mavx512f -ggdb -Wall -mtune=generic MPICHLIB_FCFLAGS=-O3 -Wall -ggdb  -msse2 -msse4.2 -mcrc32 -mavx512f -ggdb -Wall -mtune=generic MPICHLIB_F77FLAGS=-O3 -Wall -ggdb  -msse2 -msse4.2 -mcrc32 -mavx512f -ggdb -Wall -mtune=generic MPICHLIB_LDFLAGS=-O3 -L/usr/lib64  -msse2 -msse4.2 -mcrc32 -mavx512f -mtune=generic
```

Another potential drawback is that I haven't tested the ULFM feature which were scattered throughout `MPIR_Wait*_impl/Test*_impl`.

### CODE REVIEW

**Comment by hajimefu** [comment]
Fixed a build failure in CH3.

**Comment by hajimefu** [comment]
@raffenet in my local testing I didn't see any regression in CH4/OFI and CH3/Nemesis/TCP. Could you invoke a test as well to confirm?

**Comment by raffenet** [comment]
test:jenkins/ch3/all
test:jenkins/ch4/all

**Comment by hajimefu** [comment]
Note for myself: incorporate #2761 into this PR as well.

**Comment by halimamer** [comment]
it's hard to follow this PR because of the large number of whitespace changes that obfuscate the true changes. @hajimefu , could you remove the whitespace changes so that I take a look again?

**Comment by hajimefu** [comment]
Just updated the branch to
1. rebase onto the latest master
2. incorporate changes happened in #2761

@halimamer Whitespace changes are only in "Fix coding style of MPI_Test*/Wait* functions". Does it help if you ignore this last patch?

**Comment by halimamer** [comment]
yeah, I can follow better the changes of individual commits

**Comment by hajimefu** [comment]
Discussed this with @halimamer and his concern was the performance degradation. Added a patch to inline post-processing of `MPI_Waitall` (`mpi: inline waitall post-processing `). Same technique could be applied to other functions too but for now I'm just looking at the performance of it.

```
Without MPI_STATUES_IGNORE
window size  64    8192
------------------------
master	900839	751073
this PR	913105	731721

With MPI_STATUES_IGNORE
window size  64    8192
------------------------
master	927490	782224
this PR	937883	758606
```
Interestingly, now this refactoring beats the master in 64 windows (=64 requests to wait at once), but the master still wins in 8192. This trend seems consistent across ICC (18.0.1)/GCC (6.3.1), KNL/Skylake-SP, and LTO/non-LTO builds. (Haven't tried ICC LTO (IPO)). Numbers here are from ICC, non-LTO builds. Configurations are the same as above, except that now ICC is 18.0.1 and baseline is the master as of this week.

How do we proceed? Do we keep looking at this performance difference for 8192 windows? @halimamer @raffenet

**Comment by pavanbalaji** [comment]
The important question here is *why* one is better than the other.  Aren't both schemes essentially equivalent when the device does not do any special override?

**Comment by hajimefu** [comment]
They are equivalent in a logic but implementations are slightly different. Currently `MPIR_Waitall_impl` essentially looks like this:
```C
MPIR_Wait_impl(MPI_Request array_of_requests[])
{
    /* 1. handle translation */
    foreach (array_of_requests) {
        array_of_request_ptrs[i] = get_ptr(array_of_requests[i]);
    }

    foreach (array_of_request_ptrs) {
        /* 2. making progress */
        while (!complete(array_of_request_ptrs[i]))
            MPID_Progress_wait();
        /* 3. post prosessing */
        array_of_statues[i].status = array_of_request_ptrs[i].status;
        MPL_free(array_of_request_ptrs[i]);
        array_of_requests[i] = MPI_REQUEST_NULL;
    }
}

MPI_Waitall(MPI_Request array_of_requests[])
{
    MPIR_Waitall_impl(array_of_requests);
}
```
In this refactored code, `MPIR_Wait_impl (=MPID_Wait)` is only responsible for "2. making progress" part. So the entire code looks like this:
```C
MPIR_Waitall_impl(MPIR_Request array_of_request_ptrs[])
{
    foreach (array_of_request_ptrs) {
        /* 2. making progress */
        while (!complete(array_of_request_ptrs[i]))
            MPID_Progress_wait();
    }
}

MPI_Waitall(MPI_Request array_of_requests[])
{
    /* 1. handle translation */
    foreach (array_of_requests) {
        array_of_request_ptrs[i] = get_ptr(array_of_requests[i]);
    }

    /* 2. making progress */
    MPIR_Waitall_impl(array_of_request_ptrs);

     /* 3. post prosessing */
    foreach (array_of_requests) {
        array_of_statues[i].status = array_of_request_ptrs[i].status;
        MPL_free(array_of_request_ptrs[i]);
        array_of_requests[i] = MPI_REQUEST_NULL;
    }
}
```
So the current code has two loops over an array of requests, whereas the refactored one has three.

**Comment by pavanbalaji** [comment]
@hajimefu we'll need more analysis than that.  Are using three loops instead of two causing more cache misses or some other issue that's creating worse performance?  There's something more detailed that's happening and without a deep analysis to identify the problem, we won't know what to fix.

**Comment by hajimefu** [comment]
@pavanbalaji Makes sense. I'll get back to you with more deeper analysis.

**Comment by danghvu** [comment]
Here are some PAPI counters: instructions count (`TOT_INS`) and L1 data cache miss (`L1_DCM`) per `MPI_Waitall` that I have profiled for this benchmark in our KNL cluster. Ran with windows = 8192, and did not modify to ignore status:

```
master branch
size | MB/s | Messages/s | TOT_INS | L1_DCM
1 | 0.78 | 779323.83 | 1163436.04 | 2781.15
2 | 1.56 | 780292.07 | 2326846.47 | 5629.61
4 | 3.15 | 786695.38 | 3490256.54 | 8484.65
8 | 6.34 | 791960.4  | 4653666.75 | 11321.55

this pull request
size | MB/s | Messages/s | TOT_INS | L1_DCM
1 | 0.77 | 772205.74 | 1204033.79 | 5069.38
2 | 1.53 | 766035.08 | 2408041.65 | 10116
4 | 2.98 | 746067.09 | 3612049.53 | 15103.13
8 | 6.1  | 761932.77 | 4816057.53 | 20157.45
```

L1 misses for this pull request is ~2x the master branch.
So right now my guess is that the original code has better temporal locality i.e. when a request is completed, if it is freed/post-processed immediately (like the original) then you can save a miss since that request is still in cache.

**Comment by hajimefu** [comment]
@danghvu Thank you for taking a look at this. Decoupling request completion and post processing was one of the decisions we made to clearly separate responsibilities for MPI layer and MPID layer. But apparently that design comes up a non-negligible overhead.

How does it look like then the window size is 64?

**Comment by danghvu** [comment]
@hajimefu The performance impact and profiling differences are small when there are lesser requests -- the variations are high too (up to 4096 there is almost no difference -- more sample numbers below). 

I have a PR blocked depending on this PR. I'm not sure if it's worth to try something else or just to ignore this (because realistically who would MPI_Waitall on 8192 requests?). Anyway, some idea:
 - processing the list of requests in batch of 64 or less ?
 - move more responsibility to the MPID_ layer e.g .. MPIR_Request_complete/request_complete_fastpath/setting the status .. to lessen the impact ?

```
window = 64
<master>
Size | MB/s | Messages/s | TOT_INS | L1_DCM
1 | 0.87 | 871996.45 | 9723.78 | 19.19
2 | 1.78 | 889660.88 | 19420.28 | 37.24
4 | 3.75 | 937226.57 | 29116.80 | 55.60
8 | 7.31 | 913461.12 | 38813.30 | 73.86

<pr>
Size | MB/s | Messages/s | TOT_INS | L1_DCM
1 | 0.89 | 886386.69 | 9940.47 | 23.20
2 | 1.81 | 904826.96 | 19853.69 | 46.15
4 | 3.67 | 917355.87 | 29766.89 | 67.70
8 | 7.30 | 911918.77 | 39680.12 | 88.24

window = 2048
<master>
Size | MB/s | Messages/s | TOT_INS | L1_DCM
-- | -- | -- | -- | --
1 | 0.86 | 857426.33 | 297268.29 | 676.58
2 | 1.75 | 876395.84 | 594510.29 | 1356.71
4 | 3.55 | 887787.76 | 891752.26 | 2034.07
8 | 7.02 | 877305.98 | 1188994.27 | 2702.47

<pr>
Size | MB/s | Messages/s | TOT_INS | L1_DCM
1 | 0.87 | 865697.80 | 291598.10 | 694.26
2 | 1.73 | 866040.14 | 583169.71 | 1387.30
4 | 3.48 | 870861.11 | 874741.23 | 2081.12
8 | 6.99 | 874190.00 | 1166312.68 | 2771.62

window = 4096
<master>
Size | MB/s | Messages/s | TOT_INS | L1_DCM
1 | 0.83 | 832321.17 | 582197.69 | 1451.05
2 | 1.63 | 815680.82 | 1164369.77 | 2884.98
4 | 3.31 | 827535.80 | 1746542.11 | 4331.54
8 | 6.83 | 853421.68 | 2328714.18 | 5791.01
<pr>
Size | MB/s | Messages/s | TOT_INS | L1_DCM
1 | 0.83 | 831275.38 | 593499.86 | 1952.38
2 | 1.67 | 836159.50 | 1186974.44 | 3907.44
4 | 3.3 | 825739.75 | 1780449.68 | 5880.12
8 | 6.55 | 818752.92 | 2373924.36 | 7837.51


```

**Comment by pavanbalaji** [comment]
@danghvu we shouldn't ignore the performance difference, even though you might be eager to get to the other PR.  If cache misses is indeed the reason (which it seems to be based on your analysis), then batching requests, like you suggested, seems like the right solution.  Can you fixup this PR to do that?

**Comment by danghvu** [comment]
@pavanbalaji sure, I'm forking this PR.

**Comment by pavanbalaji** [comment]
This PR is superseded by PR#2957
